### PR TITLE
Callbacks

### LINKED ISSUE #3: Send snapshots and logpoints ASAP
Ideally, we'd like to see the snapshot and logpoint data as soon as it's available.

We can register a callback function when registering a snapshot/logpoint and let that callback immediately report results. As a side benefit, we then wouldn't need to keep a copy of the local variables until the end of the request.

### PR DESCRIPTION
**Pull request by chingor13**
When registering a snapshot or breakpoint, you can provide an optional `callback` option which is a PHP callable type. Rather than storing the snapshot or log message for retrieval at the end of the request, the callback is invoked.

The snapshot callback receives the snapshot array of data. The logpoint callback receives the log level, the log message, and an array of context which is compatible with a PSR-3 logger's implementation of `log($level, $message, array $context = [])`.

The callback allows us to send the data back to the server ASAP, rather than waiting until the end of the request. We will also not have to manage cleaning up memory at the end of the request.

We may decide to make this required rather than allow collection for the end of the request.

Fixes #3

### CODE REVIEW

File: stackdriver_debugger_logpoint.c [hunk 1]
```diff
@@ -151,14 +176,21 @@ void evaluate_logpoint(zend_execute_data *execute_data, stackdriver_debugger_log
     }
     ZVAL_STR(&message->message, m);
 
-    zend_hash_next_index_insert_ptr(STACKDRIVER_DEBUGGER_G(collected_messages), message);
+    if (logpoint->callback) {
+        handle_message_callback(logpoint->callback, message);
```
Comment (tmatsuo) [inline stackdriver_debugger_logpoint.c]: What happens if the callback fails?

File: stackdriver_debugger_snapshot.c [hunk 2]
```diff
@@ -421,7 +458,11 @@ void evaluate_snapshot(zend_execute_data *execute_data, stackdriver_debugger_sna
     capture_expressions(execute_data, snapshot);
 
     /* record as collected */
-    zend_hash_update_ptr(STACKDRIVER_DEBUGGER_G(collected_snapshots_by_id), snapshot->id, snapshot);
+    if (snapshot->callback) {
+        handle_snapshot_callback(snapshot->callback, snapshot);
```
Comment (tmatsuo) [inline stackdriver_debugger_snapshot.c]: ditto

**Review by tmatsuo — COMMENTED** [review]
LGTM with just a nit.

Does it make sense to emit a warning or something upon callback failure?

File: stackdriver_debugger_logpoint.c (hunk 1 above)
Comment (chingor13) [reply stackdriver_debugger_logpoint.c]: Makes sense, I'll add a test and have it emit a warning message (and clear the exception).

**Review by tmatsuo — APPROVED** [review]

**Pull request merged**